File: week2/comparison.py
import sys
import numpy as np
import time

def array_generator(n):
    array = np.random.rand(n,n)
    return array


def product_py(A,B,n):
    C = [[0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            C[i][j] = sum(A[i][k]*B[k][j] for k in range(n))
    return C


def product_np(A,B):
    product = np.dot(A,B)
    return product


if __name__ == "__main__":

    args = sys.argv
    if len(args) != 2:
        print("usage: python comparison.py <number>",file = sys.stderr)
        sys.exit(1)
    if not args[1].isdigit():
        print("invalid character detected",file = sys.stderr)
        sys.exit(1)
    n = int(args[1])

    A = array_generator(n)
    B = array_generator(n)
    t1 = time.time()
    C = product_py(A,B,n)
    t2 = time.time()
    D = product_np(A,B)
    t3 = time.time()

    path1 = "time_py.csv"
    path2 = "time_np.csv"
    with open (path1, mode="a") as f:
        f.write("%d  %lf\n" %(n, (t2-t1)*10**6))
    with open (path2, mode="a") as f:
        f.write("%d  %lf\n" %(n, (t3-t2)*10**6))

File: week2/comparison.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int array_generator(int n, float A[n][n]){

    for (int i=0; i<n; i++){
        for (int j=0; j<n; j++){
            A[i][j] = rand()*1.0/RAND_MAX;
        }
    }
    return 0;
}



int main (int argc, char** argv){

    if (argc != 2) {
        fprintf(stderr, "usage: ./comparison <number>\n");
        exit(1);
    }
    int n = atoi(argv[1]);
    if (n == 0) {
        fprintf(stderr, "error: invalid character detected\n");
        exit(1);
    }

    srand(time(NULL));

    float A[n][n];
    float B[n][n];
    array_generator(n, A);
    array_generator(n, B);
    for (int i=0; i<n; i++){
        for (int j=0; j<n; j++){
            printf("%f ",A[i][j]);
        }
        printf("\n");
    }
    for (int i=0; i<n; i++){
        for (int j=0; j<n; j++){
            printf("%f ",B[i][j]);
        }
        printf("\n");
    }

}